bob-access/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+        }
+    }
+
+    #[must_use]
+    pub fn with_authenticator(mut self, authenticator: A) -> Self {
+        self.authenticator = Some(authenticator);
+        self
+    }
+}
+
+impl<S, A> Layer<S> for AccessControlLayer<A>
+where
+    A: Clone,
+{
+    type Service = AccessControlService<A, S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        AccessControlService {
+            authenticator: self.authenticator.clone().unwrap(),
+            service,
+        }
+    }
+}
+
+type ServiceFuture<R, E> = Pin<Box<dyn Future<Output = Result<R, E>> + Send>>;
+
+impl<A, S, Request> Service<Request> for AccessControlService<A, S>
+where
+    A: Authenticator,
+    S: Service<Request>,
+    S::Error: Into<Box<dyn StdError + Send + Sync>> + 'static + Debug,
+    S::Response: Send + 'static,
+    S::Future: Send + 'static,
+    Request: Extractor + GetGrpcPermissions,
+{
+    type Response = S::Response;
+
+    type Error = Box<dyn StdError + Send + Sync>;
+
+    type Future = ServiceFuture<Self::Response, Self::Error>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx).map_err(|e| e.into())
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        debug!("request received");
+        let error = match req.extract() {
+            Ok(credentials) => {
+                debug!("credentials: {:#?}", credentials);
+                match self.authenticator.check_credentials(credentials) {
+                    Ok(permissions) => {
+                        if let Some(required_permissions) = req.get_grpc_permissions() {
+                            debug!("request requires: {}", required_permissions);
+                            if permissions.contains(required_permissions) {
+                                debug!("permissions granted");
+                                return Box::pin(self.service.call(req).map_err(|e| e.into()));
+                            } else {
+                                debug!("permissions denied");
+                                Error::PermissionDenied
+                            }
+                        } else {
+                            Error::NotGrpcRequest
+                        }
+                    }
+                    Err(e) => {
+                        warn!("Wrong credentials: {:?}", e);
+                        e
+                    }
+                }
+            }
+            Err(e) => e,
+        };
+        futures::future::err(Box::new(error) as Self::Error).boxed()
+    }
+}
+
+impl<A, S> NamedService for AccessControlService<A, S>
+where
+    S: NamedService,
+{
+    const NAME: &'static str = S::NAME;
+}
+
+pub async fn handle_auth_error(err: BoxError) -> (StatusCode, String) {
+    error!("{}", err);
+    if let Ok(err) = err.downcast::<Error>() {
+        match err.as_ref() {
+            Error::InvalidToken(_) => (StatusCode::FORBIDDEN, "Invalid token.".into()),
```

Comment:
Why there are a two places where `Error` is converted to `StatusCode` (another in [bob-access/src/error.rs](https://github.com/qoollo/bob/pull/217/files#diff-27f4a7db5d5eb59894fa6bb070daf5134db6529825c3f0edf9cd90e78a6cd6a6))?